module/zfs/mmp.c

```diff
@@ -0,0 +1,376 @@
+	mmp_ub.ub_timestamp = gethrestime_sec();
+	mmp_ub.ub_magic = MMP_MAGIC;
+	mmp_ub.ub_mmp_delay = mmp->mmp_delay;
+	vd->vdev_mmp_pending = gethrtime();
+	mutex_exit(&mmp->mmp_io_lock);
+
+	zio_t *zio  = zio_root(spa, NULL, NULL, flags);
+	abd_t *ub_abd = abd_alloc_for_io(VDEV_UBERBLOCK_SIZE(vd), B_TRUE);
+	abd_zero(ub_abd, VDEV_UBERBLOCK_SIZE(vd));
+	abd_copy_from_buf(ub_abd, &mmp_ub, sizeof (uberblock_t));
+
+	vdev_mmp_write(zio, vd, spa_get_random(VDEV_LABELS),
+	    spa_get_random(MMP_BLOCKS_PER_LABEL), ub_abd,
+	    mmp_write_done, mmp, flags | ZIO_FLAG_DONT_PROPAGATE);
+	zio_nowait(zio);
+
+	abd_free(ub_abd);
+}
+
+static void
+mmp_thread(dsl_pool_t *dp)
+{
+	mmp_thread_state_t *mmp = &dp->dp_mmp;
+	spa_t *spa = dp->dp_spa;
+	uint64_t last_zfs_mmp_interval = zfs_mmp_interval;
+	boolean_t last_spa_suspended = spa_suspended(spa);
+	callb_cpr_t cpr;
+
+	mmp_thread_enter(mmp, &cpr);
+
+	mmp->mmp_last_write = gethrtime();
+	dprintf("mmp_thread entered: pool %p\n", dp);
+
+	for (;;) {
+		/* for stable values within an iteration */
+		uint64_t mmp_fail_intervals = zfs_mmp_fail_intervals;
+		uint64_t mmp_interval = zfs_mmp_interval;
+		boolean_t suspended = spa_suspended(spa);
+		hrtime_t start, next_time;
+		hrtime_t max_fail_ns = (MSEC2NSEC(mmp_interval) *
+		    mmp_fail_intervals);
+
+		if (mmp->mmp_thread_exiting)
+			goto cleanup_and_exit;
+
+		start = gethrtime();
+		if (mmp_interval)
+			next_time = start + MSEC2NSEC(mmp_interval) /
+			    vdev_count_leaves(dp->dp_spa);
+		else
+			next_time = start + MSEC2NSEC(MMP_DEFAULT_INTERVAL);
+		/*
+		 * When MMP goes off => on, or spa goes suspended =>
+		 * !suspended, we know no writes occurred recently.  We
+		 * update mmp_last_write to give us some time to try.
+		 */
+		if ((!last_zfs_mmp_interval && mmp_interval) ||
+		    (last_spa_suspended && !suspended)) {
+			dprintf("mmp_thread transtion: pool %s last interval "
+			    "%lu interval %lu last suspended %d suspended %d\n",
+			    spa->spa_name, last_zfs_mmp_interval, mmp_interval,
+			    last_spa_suspended, suspended);
+			mutex_enter(&mmp->mmp_io_lock);
+			mmp->mmp_last_write = gethrtime();
+			mutex_exit(&mmp->mmp_io_lock);
+		} else if (last_zfs_mmp_interval && !mmp_interval) {
+			mutex_enter(&mmp->mmp_io_lock);
+			mmp->mmp_delay = 0;
+			mutex_exit(&mmp->mmp_io_lock);
+		}
+		last_zfs_mmp_interval = mmp_interval;
+		last_spa_suspended = suspended;
+
+		/*
+		 * Check after the transition check above.
+		 */
+		if (!suspended && mmp_fail_intervals && mmp_interval &&
+		    (start - mmp->mmp_last_write) > max_fail_ns) {
+			dprintf("mmp suspending pool: pool %s "
+			    "zfs_mmp_interval %lu zfs_mmp_fail_intervals %lu "
+			    "mmp_last_write %llu now %llu\n", spa->spa_name,
+			    mmp_interval, mmp_fail_intervals,
+			    mmp->mmp_last_write, start);
+			zio_suspend(spa, NULL);
+		}
+
+		if (zfs_mmp_interval) {
+			spa_config_enter(spa, SCL_STATE, FTAG, RW_READER);
+			mmp_write_uberblock(dp, &spa->spa_ubsync);
```

Comment:
Rather than use `spa->spa_ubsync` which can be older than the "best" on-disk uberblock since it's updated as the last thing in `spa_sync()`.  What if we added an `uberblock_t` to the `mmp_thread_state_t` structure and then updated it with a helper function after `vdev_uberblock_sync_list()` succeeds in `vdev_config_sync()`.  This way it would never reflect a non-op txg and it would be close this update race.

```c
void
mmp_update_uberblock(dsl_pool_t *dp, uberblock_t *ub)
{
        mmp_thread_state_t *mmp = &dp->dp_mmp;

        mutex_enter(&mmp->mmp_io_lock);
        mmp->mmp_ub = *ub;
        mmp->mmp_ub.ub_timestamp = gethrestime_sec();
        mutex_exit(&mmp->mmp_io_lock);
}
```